### PR TITLE
VAULT-15547 Agent/proxy decoupling, take two

### PR DESCRIPTION
**Pull request by VioletHynes**
A follow-up to pull request #20548 , with the following changes:
- Warnings for an Agent that's using the API proxy
- Refactor of some shared code (Persistent Cache creation and auto-auth method creation), to have less duplication and a cleaner Run function
- Additional proxy tests (mostly taken from Agent tests)
- Tests for the user agent functions

Most of the size of this PR is the copied tests. Saliently, the two functions in `helpers.go` are the ones I've used to refactor the two Run functions.

Note: the deprecations check failed for client.RawRequest, but I believe in this case, RawRequest is correct, as there is no higher level function which makes sense here. There are a few places in the code we need to use RawRequest (e.g. in the proxy's Send function), despite, in general, we should not prefer it.

### CODE REVIEW

**Review by marcboudreau — CHANGES_REQUESTED** [review]

File: command/agentproxyshared/cache/lease_cache.go [hunk 1]
```diff
@@ -174,6 +174,12 @@ func (c *LeaseCache) SetPersistentStorage(storageIn *cacheboltdb.BoltStorage) {
 	c.ps = storageIn
 }
 
+// GetPersistentStorage is a setter for the persistent storage field in
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agentproxyshared/cache/lease_cache.go:177]: Good catch!

File: command/agentproxyshared/helpers.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	case "approle":
+		return approle.NewApproleAuthMethod(authConfig)
+	case "oci":
+		return oci.NewOCIAuthMethod(authConfig, vaultAddress)
+	case "token_file":
+		return token_file.NewTokenFileAuthMethod(authConfig)
+	case "pcf": // Deprecated.
+		return cf.NewCFAuthMethod(authConfig)
+	default:
+		return nil, errors.New(fmt.Sprintf("Unknown auth method %q", autoAuthMethodType))
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agentproxyshared/helpers.go:64]: It should! Good catch :)

File: command/agent.go [hunk 3]
```diff
@@ -535,148 +499,13 @@ func (c *AgentCommand) Run(args []string) int {
...
-				if err := ps.StoreRetrievalToken(token); err != nil {
-					c.UI.Error(fmt.Sprintf("Error setting key in persistent cache: %v", err))
-					return 1
-				}
-
-				defer ps.Close()
-				leaseCache.SetPersistentStorage(ps)
-			}
+			previousToken = oldToken
+			defer deferFunc()
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agent.go:508]: There is no case where deferFunc is nil and we don't return an error, but good call-out: I'll make sure to wrap this in a nil check to protect against future potential issues :)

File: command/agentproxyshared/helpers.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+				UI.Warn(fmt.Sprintf("failed to close persistent cache file: %s", err))
+			}
+			dbFile := filepath.Join(persistConfig.Path, cacheboltdb.DatabaseFileName)
+			if err := os.Remove(dbFile); err != nil {
+				UI.Error(fmt.Sprintf("failed to remove persistent storage file %s: %s", dbFile, err))
+				if persistConfig.ExitOnErr {
+					return nil, "", fmt.Errorf("exiting with error as exit_on_err is set to true")
+				}
+			}
+			return func() error { return nil }, previousToken, nil
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agentproxyshared/helpers.go:184]: Yeah, this is a good call-out. I think returning nil here is a lot more elegant so I'll update appropriately

File: command/agentproxyshared/helpers_test.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+	if leaseCache.GetPersistentStorage() == nil {
+		t.Fatal("persistent storage was not added")
+	}
+
+	if token != "" {
+		t.Fatal("expected token to be empty")
+	}
+
+	if deferFunc == nil {
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agentproxyshared/helpers_test.go:91]: We actually should be good due to the code-path we're taking, on this particular test. We only don't defer on a non-error run if the database file already exists (and certain settings are set). I think if/when we modify this functionality a bit more, we should look into adding a few more tests, though.

File: command/proxy.go [hunk 6]
```diff
@@ -497,148 +452,13 @@ func (c *ProxyCommand) Run(args []string) int {
...
-				if err := ps.StoreRetrievalToken(token); err != nil {
-					c.UI.Error(fmt.Sprintf("Error setting key in persistent cache: %v", err))
-					return 1
-				}
-
-				defer ps.Close()
-				leaseCache.SetPersistentStorage(ps)
-			}
+			previousToken = oldToken
+			defer deferFunc()
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/proxy.go:461]: Good call, thanks!

File: helper/useragent/useragent.go [hunk 7]
```diff
@@ -75,10 +75,18 @@ func AgentAutoAuthString() string {
...
 //
-// e.g. Vault Proxy API Proxy/0.10.4 (+https://www.vaultproject.io/; go1.10.1)
+// e.g. Vault Proxy/0.10.4 (+https://www.vaultproject.io/; go1.10.1)
 func ProxyString() string {
+	return fmt.Sprintf("Vault Proxy/%s (+%s; %s)",
+		versionFunc(), projectURL, rt)
+}
+
+// ProxyAPIProxyString returns the consistent user-agent string for Vault Proxy API Proxying.
+//
```
Comment (peteski22) [inline helper/useragent/useragent.go:87]: What do you think of ditching these extra comment-spacer lines that are in this file for funcs, I haven't really seen us use them in godocs?

File: command/agentproxyshared/helpers.go [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+			return nil, "", fmt.Errorf("error setting key in persistent cache: %v", err)
+		}
+
+		leaseCache.SetPersistentStorage(ps)
+		return ps.Close, "", nil
+	}
+}
+
+// GetServiceAccountJWT reads the service account jwt from `tokenFile`. Default is
+// the default service account file path in kubernetes.
```
Comment (peteski22) [inline command/agentproxyshared/helpers.go:217]:
```suggestion
// getServiceAccountJWT attempts to read the service account JWT from the specified token file path.
// Defaults to using the Kubernetes default service account file path if token file path is empty.
```

Casing for godoc to match unexported func (and some tweakers).

**Review by peteski22 — APPROVED** [review]
As usual some comments scattered around 😄 .

It's not blocking, but I'd love it if we could take this opportunity to clean up the extracted `Run` related methods even more.

File: command/agentproxyshared/helpers.go [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+	Path                    string `hcl:"path"`
+	KeepAfterImport         bool   `hcl:"keep_after_import"`
+	ExitOnErr               bool   `hcl:"exit_on_err"`
+	ServiceAccountTokenFile string `hcl:"service_account_token_file"`
+}
+
+// AddPersistentStorageToLeaseCache adds persistence to a lease cache, based on a given PersistConfig
+// Returns a close function to be deferred and the old token, if found, or an error
+func AddPersistentStorageToLeaseCache(leaseCache *cache.LeaseCache, UI cli.Ui, persistConfig *PersistConfig, ctx context.Context, logger log.Logger) (func() error, string, error) {
+	if persistConfig.Path == "" {
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agentproxyshared/helpers.go:80]: Yeah, this is a great point. Will do!

File: command/agentproxyshared/helpers.go [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+	Type                    string
+	Path                    string `hcl:"path"`
+	KeepAfterImport         bool   `hcl:"keep_after_import"`
+	ExitOnErr               bool   `hcl:"exit_on_err"`
+	ServiceAccountTokenFile string `hcl:"service_account_token_file"`
+}
+
+// AddPersistentStorageToLeaseCache adds persistence to a lease cache, based on a given PersistConfig
+// Returns a close function to be deferred and the old token, if found, or an error
+func AddPersistentStorageToLeaseCache(leaseCache *cache.LeaseCache, UI cli.Ui, persistConfig *PersistConfig, ctx context.Context, logger log.Logger) (func() error, string, error) {
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agentproxyshared/helpers.go:79]: Yeah, this is a good point. I think UI is potentially something I can remove, too.

I tried to keep the behaviour as exactly-matching as possible, but honestly I don't think people probably care too much about how their errors look, so I'll have a go at reducing this further.

File: command/agentproxyshared/cache/lease_cache.go [hunk 11]
```diff
@@ -174,6 +174,12 @@ func (c *LeaseCache) SetPersistentStorage(storageIn *cacheboltdb.BoltStorage) {
 	c.ps = storageIn
 }
 
+// GetPersistentStorage is a getter for the persistent storage field in
+// LeaseCache
+func (c *LeaseCache) GetPersistentStorage() *cacheboltdb.BoltStorage {
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agentproxyshared/cache/lease_cache.go:179]: Thanks! Fixed

File: command/agent/config/config.go [hunk 12]
```diff
@@ -268,6 +260,17 @@ func (c *Config) Merge(c2 *Config) *Config {
 	return result
 }
 
+// DefaultListerDefined returns true if a default listener has been defined
+// in this config
+func (c *Config) DefaultListerDefined() bool {
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agent/config/config.go:265]: `Is` reads a lot better, thanks, updated!

File: command/agentproxyshared/helpers.go (hunk 8 above)
Comment (VioletHynes) [reply command/agentproxyshared/helpers.go:217]: Thanks! Good catch :)

File: helper/useragent/useragent.go (hunk 7 above)
Comment (VioletHynes) [reply helper/useragent/useragent.go:87]: Personally, I kind of like them for this specific use-case, as it leaves a space before the description and the example. I'm happy to remove them it other folks feel strongly, though. I was mostly following precedent, and the precedent looked nice enough to me for me to continue it, hehe

File: command/agent.go [hunk 13]
```diff
@@ -352,39 +346,9 @@ func (c *AgentCommand) Run(args []string) int {
...
-		case "pcf": // Deprecated.
-			method, err = cf.NewCFAuthMethod(authConfig)
-		default:
-			c.UI.Error(fmt.Sprintf("Unknown auth method %q", config.AutoAuth.Method.Type))
-			return 1
-		}
+		method, err = agentproxyshared.GetAutoAuthMethodFromConfig(config.AutoAuth.Method.Type, authConfig, config.Vault.Address)
 		if err != nil {
-			c.UI.Error(fmt.Errorf("Error creating %s auth method: %w", config.AutoAuth.Method.Type, err).Error())
+			c.UI.Error(fmt.Sprintf("Error creating %s auth method: %v", config.AutoAuth.Method.Type, err))
```
[the remark this answers is not in the document]
Comment (marcboudreau) [reply command/agent.go:351]: I think that only works with fmt.Errorf.

**Review by marcboudreau — APPROVED** [review]

File: command/agent.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (peteski22) [reply command/agent.go:351]: Ah my bad! :) Thanks Marc.

**Pull request merged**